### DISCUSSION TITLE
Audio playback stuttering problem

### DISCUSSION THREAD

**Discussion by NeilTohno** [mpv-player/mpv#13999] [Q&A] [answered]
### Important Information

Provide following Information:
- Windows Version
Windows 10 22H2
- Source of the mpv binary
https://github.com/shinchiro/mpv-winbuild-cmake/releases/tag/20240426

I added the following code to configuration file, will this cause audio playback issues?
Thanks.

```
# Audio #

audio-buffer=0.6
ad-lavc-threads=12
ao=wasapi
audio-exclusive=yes
wasapi-exclusive-buffer=6000 #min

audio-format=float
audio-channels=stereo
audio-swresample-o=resampler=soxr,precision=33
audio-swresample-o-append=cheby=1
audio-swresample-o-append=cutoff=1.00
audio-swresample-o-append=osf=fltp

# fix bgm too loud and speech too quiet
#af=lavfi="loudnorm=I=-16:TP=-3:LRA=4"
af=lavfi="loudnorm=I=-17:TP=-1:LRA=11"
```
https://github.com/mpv-player/mpv/issues/13996#issuecomment-2079266920

**Log file**:
> [   9.881][v][ao/wasapi] Accepted as stereo s16 @ 48000hz -> stereo s16 (16/16 bits) @ 48000hz (exclusive)
> [   9.881][d][ao/wasapi] Fixing format
> [   9.881][d][ao/wasapi] IAudioClient::GetDevicePeriod
> [   9.881][v][ao/wasapi] Device period: default 10000 us, minimum 3000 us
> [   9.881][v][ao/wasapi] Requested buffer duration: 6000 us
> [   9.881][v][ao/wasapi] Trying buffer duration 6000 us
> [   9.881][d][ao/wasapi] IAudioClient::Initialize
> [   9.933][d][ao/wasapi] IAudioClient::Initialize pRenderClient
> [   9.933][d][ao/wasapi] IAudioClient::Initialize IAudioClient_SetEventHandle
> [   9.933][d][ao/wasapi] IAudioClient::Initialize IAudioClient_GetBufferSize
> [   9.933][v][ao/wasapi] Buffer frame count: 288 (6000 us)
> [   9.933][v][ao/wasapi] IAudioClock::GetFrequency gave a frequency of 48000.
> [   9.934][d][ao/wasapi] Activating pEndpointVolume interface
> [   9.935][d][ao/wasapi] IAudioEndpointVolume::QueryHardwareSupport
> [   9.935][d][ao/wasapi] Entering dispatch loop
> [   9.935][d][ao/wasapi] Init wasapi done
> [   9.935][v][ao/wasapi] device buffer: 288 samples.
> [   9.935][v][ao/wasapi] using soft-buffer of 28800 samples.
> [   9.935][i][cplayer] AO: [wasapi] 48000Hz stereo 2ch s16
> [   9.935][v][cplayer] AO: Description: Windows WASAPI audio output (event mode)
> [   9.935][v][autoconvert] inserting resampler
> [   9.935][v][swresample] format change, reinitializing resampler
> [   9.936][v][swresample] 192000Hz stereo double -> 48000Hz stereo s16
> [   9.948][v][af] [out] 48000Hz stereo 2ch s16

I found lost of notification,
> mpv.com
>          ao/wasapi: Under-run: Device delay: -1400 ns
>          ao/wasapi: Under-run: Device delay: -1800 ns
>          ao/wasapi: Under-run: Device delay: -1400 ns
>          ao/wasapi: Under-run: Device delay: -1300 ns
>          ao/wasapi: Under-run: Device delay: -1300 ns
>          ao/wasapi: Under-run: Device delay: -1700 ns
>          ao/wasapi: Under-run: Device delay: -1800 ns
>          ao/wasapi: Under-run: Device delay: -1700 ns

**Comment by ruihe774** [comment]
May I ask: what's the purpose of your settings? What is your goal? Why you have to touch the default settings?

FYI, as a normal user with normal headphones, I just use the default audio settings in mpv.

**Reply by NeilTohno** [reply]
>    fix bgm too loud and speech too quiet
> af=lavfi="loudnorm=I=-17:TP=-1:LRA=11"

Usually, I add this code in the mpv‘s configuration file, I concern about the quality of audio resampling, yeah.

**Reply by ruihe774** [reply]
> I concern about the quality of audio resampling

Please do an ABX test to convince yourself that a customized config is indeed better. Nothing is "improved" if you cannot notice a difference.

> fix bgm too loud and speech too quiet

Try `--audio-normalize-downmix` first.

**Reply by NeilTohno** [reply]
48 kHz - 192 kHz - 48 kHz
Do twice resampling, I can say, it's a real problem.

> --audio-normalize-downmix

This code is useful for converting 5.1 channels to 2.0 channels.

**Reply by ruihe774** [reply]
> This code is useful for converting 5.1 channels to 2.0 channels.

I misunderstood your use case. I thought you were downmixing.

> Do twice resampling, I can say, it's a real problem.

No, IMO.

> af=lavfi="loudnorm=I=-17:TP=-1:LRA=11"

The problem you want to solve, IIRC, is exactly replaygain aims to solve.

BTW, from ffmpeg's documentation:

> However, it is recommended to run the normalization with two passes, extracting the measured values from the first run, then using the values in a second run with linear normalization enabled. See the loudnorm filter documentation for more.

I believe it is not suitable to be used in playback.

**Reply by NeilTohno** [reply]
> I believe it is not suitable to be used in playback.

There is no better method than this, so one pass is ok.

**Reply by richardpl** [reply]
No, loudnorm in one pass is bad idea.

**Reply by NeilTohno** [reply]
> No, loudnorm in one pass is bad idea.

I don't need this option anymore,  https://github.com/NeilTohno/mpv_config2/tree/main/ffmpeg

**Comment by NeilTohno** [comment]
I will test those codes with following option, funny.

```
audio-resample-max-output-size=100
msg-level=all=status,ao/wasapi=trace
```

[mpv_f.zip](https://github.com/mpv-player/mpv/files/15159062/mpv_f.zip)

**Comment by NeilTohno** [comment]
This should be ok, but problem happens again.

I'm experiencing audio popping and lagging issues when playing video files. Music files play back without any problems.

```
interpolation=yes
interpolation-threshold=-1
# autosync=30
framedrop=vo
video-sync=display-vdrop

```

**Comment by NeilTohno** [comment, marked as answer]
ao=wasapi
audio-exclusive=yes
wasapi-exclusive-buffer=21334  #(1024 samples)
